homeassistant/components/mqtt/service.py

```diff
@@ -0,0 +1,341 @@
+    """Metadata for services."""
+
+    name: str
+    type: str
+    description: str
+    required: bool
+    multiple: bool
+    custom_value: bool
+    example: str
+    selector: dict[str, SelectSelectorArgs | dict[str, Any] | None]
+    validator: Any
+
+
+ARG_TYPE_METADATA = {
+    UserServiceArgType.BOOL: ServiceArgMetadata(
+        validator=cv.boolean,
+        example="False",
+        selector={"boolean": None},
+    ),
+    UserServiceArgType.INT: ServiceArgMetadata(
+        validator=vol.Coerce(int),
+        example="42",
+        selector={"number": {ATTR_MODE: "box"}},
+    ),
+    UserServiceArgType.FLOAT: ServiceArgMetadata(
+        validator=vol.Coerce(float),
+        example="12.3",
+        selector={"number": {ATTR_MODE: "box", "step": 1e-3}},
+    ),
+    UserServiceArgType.STRING: ServiceArgMetadata(
+        validator=cv.string,
+        example="Example text",
+        selector={"text": None},
+    ),
+    UserServiceArgType.SELECT: ServiceArgMetadata(
+        validator=cv.ensure_list,
+        example="Example text",
+        selector={"select": SelectSelectorArgs()},
+    ),
+}
+
+
+def validate_options(
+    options: list[Any],
+) -> Sequence[str] | Sequence[SelectOptionDict]:
+    """Validate selector select options."""
+    if not options:
+        raise vol.Invalid("Required options are missing")
+    if isinstance(options[0], str):
+        return options
+
+    schema = vol.Schema(
+        {vol.Required("value"): cv.string, vol.Required("label"): cv.string}
+    )
+    return [
+        SelectOptionDict(value=option["value"], label=option["label"])
+        for option in options
+        if schema(option)
+    ]
+
+
+SERVICE_ARG_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_NAME): vol.All(cv.string, vol.NotIn(["dump", "publish"])),
+        vol.Required(CONF_TYPE): vol.In(
+            [arg_type.value for arg_type in UserServiceArgType]
+        ),
+        vol.Optional(CONF_DESCRIPTION): cv.string,
+        vol.Optional(CONF_EXAMPLE): cv.string,
+        vol.Optional(CONF_OPTIONS): vol.All(cv.ensure_list, validate_options),
+        vol.Optional(CONF_MULTIPLE): cv.boolean,
+        vol.Optional(CONF_CUSTOM_VALUE): cv.boolean,
+        vol.Optional(CONF_REQUIRED): cv.boolean,
+        vol.Optional(CONF_EXCLUSIVE): cv.string,
+    }
+)
+PLATFORM_SCHEMA = MQTT_BASE_SCHEMA.extend(
+    {
+        vol.Required(CONF_COMMAND_TOPIC): valid_publish_topic,
+        vol.Required(CONF_COMMAND_TEMPLATE): cv.template,
+        vol.Optional(CONF_NAME): cv.string,
+        vol.Optional(CONF_DESCRIPTION): cv.string,
+        vol.Optional(CONF_SCHEMA): vol.All(
+            [SERVICE_ARG_SCHEMA],
+            cv.ensure_list,
+        ),
+        vol.Optional(CONF_RETAIN, default=DEFAULT_RETAIN): cv.boolean,
+    },
+    extra=vol.REMOVE_EXTRA,
+)
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> None:
+    """Set up MQTT tag dynamically through MQTT discovery."""
```

Comment:
Nit: 'tag' comment copy and pasted from tag.py?

Reply:
Thnx